Generate /etc/passwd and /etc/group statically

rickynils opened the pull request:
This is a rather large commit that switches user/group creation from using
useradd/groupadd on activation to just generating the contents of /etc/passwd
and /etc/group, and then on activation merging the generated files with the
files that exist in the system. This makes the user activation process much
cleaner, in my opinion.

The users.extraUsers.<user>.uid and users.extraGroups.<group>.gid must all be
properly defined (if <user>.createUser is true, which it is by default). My
pull request adds a lot of uids/gids to config.ids to solve this problem for
existing nixos services, but there might be configurations that break because
this change. However, this will be discovered during the build.

Option changes introduced by this commit:
- Remove the options <user>.isSystemUser and <user>.isAlias since
  they don't make sense when generating /etc/passwd statically.
- Add <group>.members as a complement to <user>.extraGroups.
- Add <user>.passwordFile for setting a user's password from an encrypted
  (shadow-style) file.
- Add users.mutableUsers which is true by default. This means you can keep
  managing your users as previously, by using useradd/groupadd manually. This is
  accomplished by merging the generated passwd/group file with the existing files
  in /etc on system activation. The merging of the files is simplistic. It just
  looks at the user/group names. If a user/group exists both on the system and
  in the generated files, the system entry will be kept un-changed and the
  generated entries will be ignored. The merging itself is performed with the
  help of vipw/vigr to properly lock the account files during edit.
  If mutableUsers is set to false, the generated passwd and group files will not
  be merged with the system files on activation. Instead they will simply replace
  the system files, and overwrite any changes done on the running system. The
  same logic holds for user password, if the <user>.password or
  <user>.passwordFile options are used. If mutableUsers is false, password will
  simply be replaced on activation. If true, the initial user passwords will be
  set according to the configuration, but existing passwords will not be touched.

I have tested this on a couple of different systems and it seems to work fine
so far. If you think this is a good idea, please test it. This way of adding
local users has been discussed in issue #103 (and this commit solves that
issue).

rickynils committed:
Generate /etc/passwd and /etc/group at build time

This is a rather large commit that switches user/group creation from using
useradd/groupadd on activation to just generating the contents of /etc/passwd
and /etc/group, and then on activation merging the generated files with the
files that exist in the system. This makes the user activation process much
cleaner, in my opinion.

The users.extraUsers.<user>.uid and users.extraGroups.<group>.gid must all be
properly defined (if <user>.createUser is true, which it is by default). My
pull request adds a lot of uids/gids to config.ids to solve this problem for
existing nixos services, but there might be configurations that break because
this change. However, this will be discovered during the build.

Option changes introduced by this commit:

* Remove the options <user>.isSystemUser and <user>.isAlias since
they don't make sense when generating /etc/passwd statically.

* Add <group>.members as a complement to <user>.extraGroups.

* Add <user>.passwordFile for setting a user's password from an encrypted
(shadow-style) file.

* Add users.mutableUsers which is true by default. This means you can keep
managing your users as previously, by using useradd/groupadd manually. This is
accomplished by merging the generated passwd/group file with the existing files
in /etc on system activation. The merging of the files is simplistic. It just
looks at the user/group names. If a user/group exists both on the system and
in the generated files, the system entry will be kept un-changed and the
generated entries will be ignored. The merging itself is performed with the
help of vipw/vigr to properly lock the account files during edit.
If mutableUsers is set to false, the generated passwd and group files will not
be merged with the system files on activation. Instead they will simply replace
the system files, and overwrite any changes done on the running system. The
same logic holds for user password, if the <user>.password or
<user>.passwordFile options are used. If mutableUsers is false, password will
simply be replaced on activation. If true, the initial user passwords will be
set according to the configuration, but existing passwords will not be touched.

I have tested this on a couple of different systems and it seems to work fine
so far. If you think this is a good idea, please test it. This way of adding
local users has been discussed in issue #103 (and this commit solves that
issue).

jcumming commented:
I like it. 

There are a couple of small things that could for a UI:
- have `useradd` issue a warning, pointing to `configuration.nix` and return an error code. 
- have `useradd` build a code snippet to put into `configuration.nix`
- have `passwd` and friends issue a warning that when the system is rebuilt, the password will be overwritten

viric commented:
On Wed, May 22, 2013 at 07:57:24AM -0700, Rickard Nilsson wrote:

> This is a rather large pull request that switches user/group creation from
> using `useradd` to just generating the contents of `/etc/passwd` and
> `/etc/group`. This makes the user activation process much cleaner, in my
> opinion. However, it also makes it impossible to use `useradd` manually. All
> users must be defined in `users.extraUsers`

Impossible to use useradd manually? It's quite a bad consequence, no?

offlinehacker commented:
Is this really a good plan? Will we next time switch ifconfig with our own
implementation? While the idea might be good on the first shot, it might
rather restrict you to some /etc/passwd format instead of using
system(useradd) interface which could support any kind of format
underneath(if you supposedly let's say replace this tools with your own).
It's like changing /sys something(which could change betweene kernel
versions) by hand instead of using system tools.

On Wed, May 22, 2013 at 7:48 PM, viric notifications@github.com wrote:

> On Wed, May 22, 2013 at 07:57:24AM -0700, Rickard Nilsson wrote:
> 
> > This is a rather large pull request that switches user/group creation
> > from
> > using `useradd` to just generating the contents of `/etc/passwd` and
> > `/etc/group`. This makes the user activation process much cleaner, in my
> > 
> > opinion. However, it also makes it impossible to use `useradd` manually.
> > All
> > users must be defined in `users.extraUsers`
> 
> Impossible to use useradd manually? It's quite a bad consequence, no?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/NixOS/nixos/pull/168#issuecomment-18303223
> .

rickynils commented:
@viric Using useradd manually introduces mutability. From a Nix point of view, isn't it "better" to get a predictable set of local users each time NixOS is activated? I mount my root as tmpfs, to remove as much mutable state from my systems as possible...

rickynils commented:
@offlinehacker I agree that it would be nicer to use useradd to generate the passwd file, but unfortunately it is not possible to tell useradd to modify a custom passwd file. The current way of activating the user configuration in NixOS is hacky, since it modifies an existing passwd file that can be in any state, really. The result of that modification is of course a bit undefined.

I actually think that NixOS doesn't do anything wrong if it restricts its /etc/passwd format. In NixOS, everything is ideally well-defined, and so should the passwd file be. If we want to change the passwd format, we rewrite the NixOS module. If we want to add support for several formats, we add options and modules for that.

edolstra commented:
Doesn't this break every existing NixOS system that has users added via useradd?

rickynils commented:
@edolstra Yes, it does. But we could add something like `users.mutableUsers` which defaults to true and that could merge the passwd and group files from the store with the ones in /etc, instead of (if set to false) replacing them.

offlinehacker commented:
Okay, i see the problem, i support.
On May 23, 2013 1:05 AM, "Rickard Nilsson" notifications@github.com wrote:

> @edolstra https://github.com/edolstra Yes, it does. But we could add
> something like users.mutableUsers which defaults to true and that could
> merge the passwd and group files from the store with the ones in /etc,
> instead of (if set to false) replacing them.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/NixOS/nixos/pull/168#issuecomment-18313925
> .

peti commented:
It's good to ensure that all users and groups are configured via configuration.nix -- I would like tat.

It's bad to break a running system during nixos-rebuild, though. As long as that is the case, I really don't want this patch applied.

rickynils commented:
I will add an option that disables the replacement of the `passwd` and `group` files, and set that option enabled by default in order not to break existing systems.

edolstra commented:
But in that case we still need all the script code to update passwd/group, so we don't really gain a lot. Instead we end up with two different code paths for dealing with passwd/group...

rickynils commented:
@edolstra Yes, two code paths are unavoidable, if we want to support two behaviors. But the paths should not differ too much? We can keep the generation of the files the same in both cases, and just let the "mutable" case merge the generated files with the existing ones. In the "immutable" case, we just do a replacement. Is this not acceptable?

Another, cleaner solution, would be to let NixOS ignore /etc/passwd,group completely, and instead use some additional nss module that could read the user information generated by NixOS. I looked briefly at nss_db, but it seems a bit messy. An alternative could be to hack glibc and make a libnss_files2.so that could read from /etc/passwd2,group2. But it is considerably more work...

rickynils commented:
I have updated this pull request. Now, the generated passwd/group files are merge with the existing one on activation. The merge is done with help of `vipw` and `vigr` to make sure everything is locked properly. You can set users.mutableUsers to false if you want to skip merging and just replace the existing files with the ones from the Nix store. The code paths for merging vs replacing is very similar.

The merging itself is simplistic. It only looks at user/group names when comparing. If a user/group exist both on the system and in the generated files, the system entry will be kept unchanged.

rickynils commented:
I have now set the root password to blank by default, to make my changes compliant with the old behavior. I have also removed the option `users.lockRoot` that I added. You can instead lock any user password by just setting `users.extraUsers.<user>.password = null` and `users.extraUsers.<user>.passwordFile = null` (which is the default, except for root, that now has `password = ""`

I have also tested various scenarios in VMs now, and I think I can say that it works now, and that no existing behavior should break. You can use `useradd` and `passwd` just like before. Existing, manually added users/groups/passwords will only be replaced on activation if you set `mutableUsers = false`.

@edolstra @peti @jcumming @viric @offlinehacker : Comments?

offlinehacker commented:
Just one thing. I haven't read the code exactly. But why do you have to now
set uid and gid in every service, what is stopping you with just only using
name parameter as before?

On Thu, Jul 4, 2013 at 5:06 PM, Rickard Nilsson notifications@github.comwrote:

> I have now set the root password to blank by default, to make my changes
> compliant with the old behavior. I have also removed the option
> users.lockRoot that I added. You can instead lock any user password by
> just setting users.extraUsers.<user>.password = null and users.extraUsers.<user>.passwordFile
> = null (which is the default, except for root, that now has password = ""
> 
> I have also tested various scenarios in VMs now, and I think I can say
> that it works now, and that no existing behavior should break. You can use
> useradd and passwd just like before. Existing, manually added
> users/groups/passwords will only be replaced on activation if you set mutableUsers
> = false.
> 
> @edolstra https://github.com/edolstra @peti https://github.com/peti
> @jcumming https://github.com/jcumming @viric https://github.com/viric
> @offlinehacker https://github.com/offlinehacker : Comments?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/NixOS/nixos/pull/168#issuecomment-20481780
> .

rickynils commented:
@offlinehacker the uid and gid are the numeric ids, we can't use the name for that. Before, `useradd` picked a free uid/gid itself on activation if they weren't defined in the configuration. That is not so deterministic. Note, you can still get clashes if `mutableUsers = true` and you add manually add a user with one uid, and after that configure another user (another name) with the same uid in the NixOS configuration. This could maybe be improved with better merging, but could still be clashes. `mutableUsers = false` avoids any clashes and assures the uids/gids are unique.

mornfall commented:
This would be useful, as we already ran into UID de-syncing problems with NFS3 shared across NixOS boxes. Current "pick some circumstances-dependent uid" is very nasty, impure and I would be very grateful if we could deal with it, even if the upgrade path is not the cleanest imaginable.

rickynils commented:
@edolstra Any chance for merging this? If `users.mutableUsers = false` (the default) there shouldn't be any functional differences to the way users and groups are managed today. The code path difference when setting it to true is minimal.

I have been using this on several systems for a long time now, with no apparent problems.

edolstra commented:
You mean that the default is true, right? That would be fine.

offlinehacker commented:
+1 for merging this, looks good to me
On Jul 31, 2013 4:26 PM, "Rickard Nilsson" notifications@github.com wrote:

> @edolstra https://github.com/edolstra Any chance for merging this? If users.mutableUsers
> = false (the default) there shouldn't be any functional differences to
> the way users and groups are managed today. The code path difference when
> setting it to true is minimal.
> 
> I have been using this on several systems for a long time now, with no
> apparent problems.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/NixOS/nixos/pull/168#issuecomment-21866157
> .

rickynils commented:
@edolstra Yes, I mean `users.mutableUsers = true` by default, of course.

domenkozar commented:
Then we have to maintain two ways how users/groups are generated, is this really such a benefit?

edolstra commented:
Well, the "new" way is conceptually much nicer because it fits the NixOS model very well. E.g. removing a user from `users.extraUsers` actually causes that user to disappear. Unfortunately the old way has to be supported as well in order not to break everybody's existing system.

rickynils commented:
@iElectric The two "ways" are this: At build time, generate `passwd` and `group` files from the NixOS config. On activation, if `users.mutableUsers = false`, put those two files in `/etc`, replacing any existing files. If `users.mutableUsers = true`, merge the built files with the existing ones instead.

domenkozar commented:
Understood. Doesn't seem too complicated indeed :) Let's make sure that is well-phrased in documentation string.

bjornfor commented:
@rickynils: I like this. Just a few comments.
- The commit message says mutableUsers is false by default, which is incorrect. The next sentence after that also has mutableUsers true/false mixed up (I think).
- lighttpd user and group are missing. Could you add them?

Does having no password attribute mean that users can manage their own password with the "passwd" program?
I'd really like to be able to activate a system with users.extraUsers.foo.password = "changeme" and then immediately comment out that line and let the user manage his own password. Is that already possible?

rickynils commented:
@bjornfor Sorry for the true/false confusion about mutableUsers. I corrected the texts in the pull request, the commit and the option description now.

I've added lighttpd user and group.

No password attribute means disabled password (`passwd -l`). If `mutableUsers = true`, users can manage their passwords as usual. Only on the initial user creation, the password will be set according to the configuration. If `mutableUsers = false`, a user without password attribute would have his password disabled on each activation.

Thinking about the password scenario now, I realise that passwords might need to be mutable even though the rest of the user stuff isn't mutable. I could change the meaning of "no password attribute" into "do not touch password" (along @bjornfor's suggestion). The only downside of that is that we would have no way of forcing passwords that we want to have disabled (like for system accounts) to keep being disabled. If someone sets a password for a system account, that password wouldn't be reset during activation. One way of solving that could be to somehow turn the user.password option into a tri-state thing: `do not touch password | disable password | set password` (with a default of disabling passwords, to follow the existing convention in NixOS, where a user has no home or shell if not specified). Any ideas on how that could be implemented in a nice way?

peti commented:
Suppose I want mutableUsers = false, but I don't want my user passwords spelled out in configuration.nix either. Is that possible?

Rickard Nilsson notifications@github.com wrote:

> @bjornfor Sorry for the true/false confusion about mutableUsers. I corrected the texts in the pull request, the commit and the option description now.
> 
> I've added lighttpd user and group.
> 
> No password attribute means disabled password (`passwd -l`). If `mutableUsers = true`, users can manage their passwords as usual. Only on the initial user creation, the password will be set according to the configuration. If `mutableUsers = false`, a user without password attribute would have his password disabled on each activation.
> 
> Thinking about the password scenario now, I realise that passwords might need to be mutable even though the rest of the user stuff isn't mutable. I could change the meaning of "no password attribute" into "do not touch password" (along @bjornfor's suggestion). The only downside of that is that we would have no way of forcing passwords that we want to have disabled (like for system accounts) to keep being disabled. If someone sets a password for a system account, that password wouldn't be reset during activation. One way of solving that could be to somehow turn the user.password option into a tri-state thing: `do not touch password | disable password | set password` (with a default of disabling passwords, to follow the existing convention in NixOS, where a user has no home or shell if not specified). Any ideas on how that could be implemented in a nice way?
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/NixOS/nixos/pull/168#issuecomment-21974027

rickynils commented:
@peti Yes, you can use the user.passwordFile option instead, and store your user passwords encrypted somewhere instead. That is how I handle my "real" local users.

If the "do not touch password" feature is implemented you could of course also just manage your passwords in the plain old mutable way, with `passwd` updating the shadow file. Would you prefer to have that option?

bjornfor commented:
On 2 August 2013 00:17, Rickard Nilsson notifications@github.com wrote:

> @bjornfor Sorry for the true/false confusion about mutableUsers. I corrected the texts in the pull request, the commit and the option description now.

Thanks.

> I've added lighttpd user and group.

Thanks.

> No password attribute means disabled password (passwd -l). If mutableUsers = true, users can manage their passwords as usual. Only on the initial user creation, the password will be set according to the configuration. If mutableUsers = false, a user without password attribute would have his password disabled on each activation.
> 
> Thinking about the password scenario now, I realise that passwords might need to be mutable even though the rest of the user stuff isn't mutable. I could change the meaning of "no password attribute" into "do not touch password" (along @bjornfor's suggestion). The only downside of that is that we would have no way of forcing passwords that we want to have disabled (like for system accounts) to keep being disabled. If someone sets a password for a system account, that password wouldn't be reset during activation. One way of solving that could be to somehow turn the user.password option into a tri-state thing: do not touch password | disable password | set password (with a default of disabling passwords, to follow the existing convention in NixOS, where a user has no home or shell if not specified). Any ideas on how that could be implemented in a nice way?

Isn't password = null already different from undefined password
attribute (in nix language)? If so, I don't see any issue with
implementing the tri-state thing in a pretty straight forward way:
- undefined password attribute: do not touch password (leave it as is)
- password = null: passwd --lock mode (password login inhibited)
- password = "my secret": obvious :-)

rickynils commented:
@bjornfor When I said "undefined", I actually meant "null". However, it is possible to have really undefined options by using `mkNotdef` and `isNotDef`. That works (I've just tested it), but not if we want to have `null` (locked password) as default. And I think that we really want to have locked passwords as a default, to keep system users from getting passwords at all.

rickynils commented:
@bjornfor @edolstra The nicest implementation would really be a tagged union type: `password : Locked | Not managed | File <string> | Password <string>`. Can that be represented in Nix in a simple way?

MarcWeber commented:
@peti > Don't want user passwords spelled out in configuration.nix 
The problem with not using a password file is not only configuration,nix,
but the passwords will end up in the store, won't they?
Otherwise you could just use (import ..) ..

This patch tries to automate creating such files, thus that you can place
passwords in passwords.nix and import it in configuration.nix:
https://github.com/MarcWeber/nix/blob/experimental/write-file-hashed/.topmsg
Of course because the root user rebuilds the system those automatically
generated files. Thus a script with root privileges must be run reading those contents

offlinehacker commented:
I would really like this to get merged. Current implementation has many errors like home folders not being recreated if user exists. This prevents some modules that use home folder creation capabilites to /var/someting behave correctly.

rickynils commented:
@MarcWeber @peti No, if you set `user.passwordFile` it will not end up in the store. The file will be read on system activation, so it can be put in a safe place.

rickynils commented:
@offlinehacker I'm also in favor of a merge. The only limitation I can think of is the password handling if you set `users.mutableUsers = false` (which is non-default). See my comments above about that.

edolstra commented:
We can merge this after the upcoming NixOS release.

offlinehacker commented:
It would be also nice to create homes recursively (with -p) see #262.

rickynils committed:
Create home directories with mkdir -p

offlinehacker commented:
Can you please reopen that pull request in nixpkgs(this repo is deprecated) so i can simple merge it there localy?

rickynils commented:
@offlinehacker Done: https://github.com/NixOS/nixpkgs/pull/1076

MarcWeber commented:
If the request has been moved this can be closed.
